File: Application/src/main/java/com/shareyourproxy/api/rx/command/DeleteUserGroupCommand.java
package com.shareyourproxy.api.rx.command;

import android.app.Service;
import android.os.Parcel;
import android.support.annotation.NonNull;

import com.shareyourproxy.api.domain.model.Group;
import com.shareyourproxy.api.domain.model.User;
import com.shareyourproxy.api.rx.RxUserGroupSync;
import com.shareyourproxy.api.rx.command.eventcallback.EventCallback;

import java.util.List;

/**
 * Delete a group associated with a User.
 */
public class DeleteUserGroupCommand extends BaseCommand {
    public static final Creator<DeleteUserGroupCommand> CREATOR =
        new Creator<DeleteUserGroupCommand>() {
            @Override
            public DeleteUserGroupCommand createFromParcel(Parcel in) {
                return new DeleteUserGroupCommand(in);
            }

            @Override
            public DeleteUserGroupCommand[] newArray(int size) {
                return new DeleteUserGroupCommand[size];
            }
        };
    private final static java.lang.ClassLoader CL = DeleteUserGroupCommand.class.getClassLoader();
    public final Group group;
    public final User user;

    /**
     * Public constructor.
     *
     * @param user  logged in user
     * @param group this events group
     */
    public DeleteUserGroupCommand(@NonNull User user, @NonNull Group group) {
        this.user = user;
        this.group = group;
    }

    private DeleteUserGroupCommand(Parcel in) {
        this((User) in.readValue(CL), (Group) in.readValue(CL));
    }

    @Override
    public List<EventCallback> execute(Service service) {
        return RxUserGroupSync.deleteUserGroup(service, user, group);
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeValue(user);
        dest.writeValue(group);
    }
}

File: Application/src/main/kotlin/com/shareyourproxy/api/rx/command/DeleteUserGroupCommand.kt
package com.shareyourproxy.api.rx.command

import android.app.Service
import android.os.Parcel
import android.os.Parcelable
import com.shareyourproxy.api.domain.model.Group
import com.shareyourproxy.api.domain.model.User
import com.shareyourproxy.api.rx.RxUserGroupSync.deleteUserGroup
import com.shareyourproxy.api.rx.command.eventcallback.EventCallback


/**
 * Delete a group associated with a User.
 * @param user  logged in user
 * @param group this events group
 */
internal final class DeleteUserGroupCommand(val user: User, val group: Group) : BaseCommand() {

    private constructor(parcel: Parcel) : this(parcel.readValue(CL) as User, parcel.readValue(CL) as Group)

    override fun execute(service: Service): EventCallback {
        return deleteUserGroup(service, user, group)
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeValue(user)
        dest.writeValue(group)
    }

    companion object {
        private val CL = DeleteUserGroupCommand::class.java.classLoader
        val CREATOR: Parcelable.Creator<DeleteUserGroupCommand> = object : Parcelable.Creator<DeleteUserGroupCommand> {
            override fun createFromParcel(parcel: Parcel): DeleteUserGroupCommand {
                return DeleteUserGroupCommand(parcel)
            }

            override fun newArray(size: Int): Array<DeleteUserGroupCommand?> {
                return arrayOfNulls(size)
            }
        }
    }
}
